tests/resp.py

```diff
@@ -0,0 +1,532 @@
+import itertools
+from contextlib import closing
+from types import NoneType
+from typing import Any, Generator, List, Optional, Tuple, Union
+
+CRNL = b"\r\n"
+
+
+class VerbatimStr(str):
+    """
+    A string that is encoded as a resp3 verbatim string
+    """
+
+    def __new__(cls, value: str, hint: str) -> "VerbatimStr":
+        return str.__new__(cls, value)
+
+    def __init__(self, value: str, hint: str) -> None:
+        self.hint = hint
+
+    def __repr__(self) -> str:
+        return f"VerbatimStr({super().__repr__()}, {self.hint!r})"
+
+
+class ErrorStr(str):
+    """
+    A string to be encoded as a resp3 error
+    """
+
+    def __new__(cls, code: str, value: str) -> "ErrorStr":
+        return str.__new__(cls, value)
+
+    def __init__(self, code: str, value: str) -> None:
+        self.code = code.upper()
+
+    def __repr__(self) -> str:
+        return f"ErrorString({self.code!r}, {super().__repr__()})"
+
+    def __str__(self):
+        return f"{self.code} {super().__str__()}"
+
+
+class PushData(list):
+    """
+    A special type of list indicating data from a push response
+    """
+
+    def __repr__(self) -> str:
+        return f"PushData({super().__repr__()})"
+
+
+class Attribute(dict):
+    """
+    A special type of map indicating data from a attribute response
+    """
+
+    def __repr__(self) -> str:
+        return f"Attribute({super().__repr__()})"
+
+
+class RespEncoder:
+    """
+    A class for simple RESP protocol encoder for unit tests
+    """
+
+    def __init__(
+        self, protocol: int = 2, encoding: str = "utf-8", errorhander="strict"
+    ) -> None:
+        self.protocol = protocol
+        self.encoding = encoding
+        self.errorhandler = errorhander
+
+    def apply_encoding(self, value: str) -> bytes:
+        return value.encode(self.encoding, errors=self.errorhandler)
+
+    def has_crnl(self, value: bytes) -> bool:
+        """check if either cr or nl is in the value"""
+        return b"\r" in value or b"\n" in value
+
+    def escape_crln(self, value: bytes) -> bytes:
+        """remove any cr or nl from the value"""
+        return value.replace(b"\r", b"\\r").replace(b"\n", b"\\n")
+
+    def encode(self, data: Any, hint: Optional[str] = None) -> bytes:
+        if isinstance(data, dict):
+            if self.protocol > 2:
+                code = "|" if isinstance(data, Attribute) else "%"
+                result = f"{code}{len(data)}\r\n".encode()
```

Comment:
maybe use `CRNL` variable here and in all other lines below since its already defined?

Reply:
no, CRNL is a bytes object, just a shorthand for parsing, not emitting where we use str.

Reply:
sorry, i commented on wrong line. there are still a few insurances of `b"\r\n"` in code in addition to strings

Reply:
Yes, I like it like that since it makes the code more consistent and explicit where we are writing RESP responses.